hedera-node/hedera-app/src/main/java/com/hedera/node/app/tss/TssBaseServiceImpl.java

```diff
@@ -380,12 +407,359 @@ public TssMessage getTssMessageFromBytes(Bytes wrap, TssParticipantDirectory dir
+        private final Instant consensusNow;
+        private final RosterAndTssInfo info;
+
+        public StatusChange(
+                final boolean isStakePeriodBoundary, final Instant consensusNow, final RosterAndTssInfo info) {
+            this.isStakePeriodBoundary = isStakePeriodBoundary;
+            this.info = info;
+            this.newKeyingStatus = tssStatus.tssKeyingStatus();
+            this.newRosterToKey = tssStatus.rosterToKey();
+            this.newLedgerId = tssStatus.ledgerId();
+            this.consensusNow = consensusNow;
+        }
+
+        /**
+         * Computes the new status based on the old status and the current state of the system.
+         * If needed, it schedules work to generate TSS messages and votes.
+         *
+         * @return the new status
+         */
+        public TssStatus computeNewStatus() {
+            switch (tssStatus.rosterToKey()) {
+                case NONE -> {
+                    if (isStakePeriodBoundary) {
+                        newRosterToKey = CANDIDATE_ROSTER;
+                        newKeyingStatus = TssKeyingStatus.WAITING_FOR_THRESHOLD_TSS_MESSAGES;
+                        haveSentMessageForTargetRoster = false;
+                        haveSentVoteForTargetRoster = false;
+                    }
+                }
+                case CANDIDATE_ROSTER -> {
+                    final var activeRosterHash = requireNonNull(info.activeRosterHash());
+                    final var candidateRosterHash = RosterUtils.hash(requireNonNull(info.candidateRoster()))
+                            .getBytes();
+
+                    switch (tssStatus.tssKeyingStatus()) {
+                        case KEYING_COMPLETE -> newRosterToKey = NONE;
+                        case WAITING_FOR_THRESHOLD_TSS_MESSAGES -> validateMessagesAndSubmitIfNeeded(
+                                activeRosterHash, candidateRosterHash);
+                        case WAITING_FOR_THRESHOLD_TSS_VOTES -> validateVotesAndSubmitIfNeeded(
+                                activeRosterHash, candidateRosterHash);
+                        case WAITING_FOR_ENCRYPTION_KEYS -> validateThresholdEncryptionKeysReached(info.activeRoster());
+                    }
+                }
+                case ACTIVE_ROSTER -> {
+                    requireNonNull(info.activeRosterHash());
+                    switch (tssStatus.tssKeyingStatus()) {
+                        case KEYING_COMPLETE -> newRosterToKey = NONE;
+                        case WAITING_FOR_THRESHOLD_TSS_MESSAGES -> validateMessagesAndSubmitIfNeeded(
+                                Bytes.EMPTY, info.activeRosterHash());
+                        case WAITING_FOR_THRESHOLD_TSS_VOTES -> validateVotesAndSubmitIfNeeded(
+                                Bytes.EMPTY, info.activeRosterHash());
+                        case WAITING_FOR_ENCRYPTION_KEYS -> validateThresholdEncryptionKeysReached(
+                                info.candidateRoster());
```

Comment:
And if we are keying the active roster, we should use `info.activeRoster()` here.

Reply:
yes Changed . Thanks